### PR TITLE
scripts: Exclude Rust compilation units with pahole

### LINKED ISSUE #735: Module loading fails with DEBUG_INFO_BTF_MODULES
**Description**

While testing the v5 patchset with the Arch Linux config (see below) I've stumbled upon module loading problems. I could reproduce the issue with the v5.16 series but more relevantly the v5.17 series. Both QEMU and bare metal don't want to work with it. Builds with `CONFIG_DEBUG_INFO_BTF` disabled load the modules as expected.

Stems from #695.

**Architecture(s)**
  - `x86_64`

**Toolchain versions**
  - `rustc`: 1.59.0
  - `bindgen`: 0.56.0
  - LLVM/Clang: 13.0.1
  - GCC: 11.2.0
  - QEMU: 6.2.0

<details><summary><strong>Kernel log</strong></summary>
<p>

```text
[    0.642535] BPF:     width type_id=54975 bits_offset=0
[    0.642932] BPF:  
[    0.643100] BPF: Invalid member bits_offset
[    0.643413] BPF:
```

</p>
</details>

<details><summary><strong>Kernel config</strong></summary>
<p>

```text
#
# Compile-time checks and compiler options
#
CONFIG_DEBUG_INFO=y
# CONFIG_DEBUG_INFO_REDUCED is not set
# CONFIG_DEBUG_INFO_COMPRESSED is not set
# CONFIG_DEBUG_INFO_SPLIT is not set
# CONFIG_DEBUG_INFO_DWARF_TOOLCHAIN_DEFAULT is not set
CONFIG_DEBUG_INFO_DWARF4=y
CONFIG_DEBUG_INFO_BTF=y
CONFIG_PAHOLE_HAS_SPLIT_BTF=y
CONFIG_DEBUG_INFO_BTF_MODULES=y
CONFIG_GDB_SCRIPTS=y
CONFIG_FRAME_WARN=2048
CONFIG_STRIP_ASM_SYMS=y
# CONFIG_HEADERS_INSTALL is not set
CONFIG_SECTION_MISMATCH_WARN_ONLY=y
CONFIG_STACK_VALIDATION=y
# CONFIG_DEBUG_FORCE_WEAK_PER_CPU is not set
# end of Compile-time checks and compiler options
```

</p>
</details>

### PR DESCRIPTION
**Pull request by YakoYakoYokuYoku**
Version 1.24 of pahole has the capability to exclude compilation units (CUs) of specific languages with the `--lang_exclude` option.
Rust, as of writing, is not currently supported by pahole neither BTF and if it's used with a build that has `DEBUG_INFO_BTF_MODULES` enabled it results in erroneous binaries. So it's better for pahole to exclude Rust compilation units until support for it arrives.

Fixes #735.

### CODE REVIEW

File: init/Kconfig
Comment (ojeda) [inline init/Kconfig]: Should we follow the other instances of `PAHOLE_HAS_*` and create a `PAHOLE_HAS_LANG_EXCLUDE` or similar?

i.e.

```Kconfig
config PAHOLE_HAS_SPLIT_BTF
	def_bool PAHOLE_VERSION >= 119

config PAHOLE_HAS_BTF_TAG
	def_bool PAHOLE_VERSION >= 123
	depends on CC_IS_CLANG
	help
	  Decide whether pahole emits btf_tag attributes (btf_type_tag and
	  btf_decl_tag) or not. Currently only clang compiler implements
	  these attributes, so make the config depend on CC_IS_CLANG.
```

Comment (YakoYakoYokuYoku) [reply init/Kconfig]: Damn, you are right. Ty for the pointer.

File: lib/Kconfig.debug
Comment (ojeda) [inline lib/Kconfig.debug]:
```suggestion
	  Support for the --lang_exclude flag which makes pahole exclude
```

File: lib/Kconfig.debug
Comment (ojeda) [inline lib/Kconfig.debug]: I would take the chance to explain here why we exclude them (so that people don't need to reach for the commit message).

Then in the `pahole-flags.sh` file you could add a comment referring to that explanation, since one of the other `if`s has a comment:

```sh
	# pahole 1.18 through 1.21 can't handle zero-sized per-CPU vars
	extra_paholeopt="${extra_paholeopt} --skip_encoding_btf_vars"
```

e.g.

```sh
	# See PAHOLE_HAS_LANG_EXCLUDE
	extra_paholeopt="${extra_paholeopt} --lang_exclude=rust"
```

Or maybe the other way around (i.e. the explanation in the `.sh`), though having it in the config option may be easier for people that sees the dependency on it.

File: init/Kconfig
Comment (ojeda) [reply init/Kconfig]: No worries!

File: lib/Kconfig.debug
Comment (YakoYakoYokuYoku) [reply lib/Kconfig.debug]: Had to check my grammar, and yes, this suggestion seems to be correct.

File: lib/Kconfig.debug
Comment (YakoYakoYokuYoku) [reply lib/Kconfig.debug]: Prefered to leave the explanation in the `Kconfig` file and refer to it in `pahole-flags.sh`, both to maintain things in one place and to explain things to `menuconfig` users.

File: lib/Kconfig.debug
Comment (ojeda) [reply lib/Kconfig.debug]: I am not a native speaker, so I may be wrong, but it sounded strange to me. :)

**Comment by YakoYakoYokuYoku** [comment]
To test this change (as of 459035a) the flag `--skip_encoding_btf_enum64` needs to be passed in `scripts/pahole-flags.sh`, though this change might not be necessary here as it's gonna land in upstream probably.

**Comment by ojeda** [comment]
> To test this change (as of [459035a](https://github.com/Rust-for-Linux/linux/commit/459035ab65c0ebb8d7054b24b6c00de907819eb2)) the flag `--skip_encoding_btf_enum64` needs to be passed in `scripts/pahole-flags.sh`, though this change might not be necessary here as it's gonna land in upstream probably.

Sorry, I got this swapped out from my brain -- could you please expand a bit on why the flag is needed for testing?

The PR otherwise looks good to me, though this probably should go through the BPF tree after the Rust merge. Did you mean that by "gonna land in upstream probably"?

**Comment by YakoYakoYokuYoku** [comment]
> Sorry, I got this swapped out from my brain -- could you please expand a bit on why the flag is needed for testing?
> 
> The PR otherwise looks good to me, though this probably should go through the BPF tree after the Rust merge. Did you mean that by "gonna land in upstream probably"?

There was an issue related to 64 bit enums with the stable branches of the kernel. Though it's solved for both stable and 6.0+ because with the former they've added `--skip_encoding_btf_enum64` to `scripts/pahole-flags.sh` to avoid the encoding and the latter supports those kinds of enums. In resumi, upstream issues had been fixed and there are no blockers for this.

**Comment by YakoYakoYokuYoku** [comment]
@ojeda I think this is now ready for inclusion.

**Comment by ericcurtin** [comment]
We really need this upstream if this does fix the bpf issue with Rust for Linux, is CONFIG_MODULE_ALLOW_BTF_MISMATCH=y required also or no?

**Comment by ericcurtin** [comment]
I can confirm this worked great for me, to get BPF back working with the Rust-based GPU driver on our Fedora Asahi kernel for the non-Rust parts:

https://gitlab.com/fedora-asahi/kernel-asahi/-/commit/cb4241a09a5eeb51d07455df2ca8a1a882990c6f

So...

```
Reviewed-by: Eric Curtin <ecurtin@redhat.com>
Tested-by: Eric Curtin <ecurtin@redhat.com>
```

**Comment by ericcurtin** [comment]
@marcan

**Comment by YakoYakoYokuYoku** [comment]
> We really need this upstream if this does fix the bpf issue with Rust for Linux, is `CONFIG_MODULE_ALLOW_BTF_MISMATCH=y` required also or no?

I don't think so as `CONFIG_PAHOLE_HAS_LANG_EXCLUDE=y` already suffices the condition of having `--lang_exclude=rust`.

Also I've added those tags in the commit message to credit you, @ericcurtin, so thanks to you too for testing this PR.

**Comment by ericcurtin** [comment]
> > We really need this upstream if this does fix the bpf issue with Rust for Linux, is `CONFIG_MODULE_ALLOW_BTF_MISMATCH=y` required also or no?
> 
> I don't think so as `CONFIG_PAHOLE_HAS_LANG_EXCLUDE=y` already suffices the condition of having `--lang_exclude=rust`.
> 
> Also I've added those tags in the commit message to credit you, @ericcurtin, so thanks to you too for testing this PR.

Yup you are right, tested without it, works great!

**Comment by Conan-Kudo** [comment]
Verified as well:

```
Reviewed-by: Neal Gompa <neal@gompa.dev>
Tested-by: Neal Gompa <neal@gompa.dev>
```

**Comment by YakoYakoYokuYoku** [comment]
You were added too, @Conan-Kudo.

**Review by Conan-Kudo — APPROVED** [review]

**Comment by ojeda** [comment]
Thanks everyone! The tags are very useful.

@YakoYakoYokuYoku For upstream, could you please submit the patch? If so, please Cc the BPF list & maintainers and probably the `pahole` one (@acmel) too, as well as the Rust for Linux list.

**Comment by YakoYakoYokuYoku** [comment]
[There it goes](https://lore.kernel.org/rust-for-linux/20221220203901.1333304-1-yakoyoku@gmail.com/T/#u), though I've fooled up and the message was sent twice. If I recall fixes should be sent upstream directly, right?

**Comment by ojeda** [comment]
> There it goes, though I've fooled up and the message was sent twice.

Thanks, and no worries :)

> If I recall fixes should be sent upstream directly, right?

At this point, as soon as something has all its dependencies already in mainline, it can be sent.

**Comment by YakoYakoYokuYoku** [comment]
[v2 at LKML](https://lore.kernel.org/rust-for-linux/20230108021450.120791-1-yakoyoku@gmail.com/T/#u).

**Comment by YakoYakoYokuYoku** [comment]
[v3 at LKML](https://lore.kernel.org/rust-for-linux/20230111152050.559334-1-yakoyoku@gmail.com/T/#u).

**Comment by YakoYakoYokuYoku** [comment]
@ojeda, can you merge this? This was accepted for `bpf-next` so this can be included unless we wait for it to be in `torvalds` and pulled from there.

**Comment by ojeda** [comment]
Sure (in any case, please note that the `rust` branch will eventually go away when we get it close to upstream)

**Pull request merged**
